Add Buildpack environment related content (#754)
buildpacks/docs: content/docs/for-buildpack-authors/how-to/write-buildpacks/clear-env.md

weight=99
+++

"Clearing" the buildpack environment with `clear-env` is the process of preventing end-users from customizing a buildpack's behavior through environment variables.

<!--more-->

> Buildpack authors may elect to clear user-defined environment variables when `bin/detect` and `bin/build` are executed. This is achieved by setting `clear-env` to `true` in [buildpack.toml](https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml); by default `clear-env` is set to `false`.

Author: updated

> * When `clear-env` is set to `true` for a given buildpack, the `lifecycle` will not set user-provided environment variables when running `/bin/detect` or `/bin/build` for the given buildpack.

Author: updated

weight=99
+++

> Clearing the buildpack environment `clear-env` is the process of preventing end-users from customizing a buildpack.

Reviewer: ```suggestion
"Clearing" the buildpack environment with `clear-env` is the process of preventing end-users from customizing a buildpack's behavior through environment variables.
```

Author: added

> During the `detect` and `build` phases, the lifecycle MUST provide as environment variables any user-provided files in `<platform>/env/` with environment variable names and contents matching the file names and contents.

Reviewer: The `<platform>/env/` directory is just the mechanism by which the platform can provide build-time environment variables for buildpacks to the lifecycle, so the lifecycle can set those up. Buildpack authors don't really have to care that it's called `<platform>/env/`, just that there is a way for end users to specify build-time environment variables. "Cear env" is a way for a buildpack to declare "I am not configurable by the end user". 

Note that buildpacks CANNOT declare "I am not configurable by the platform". That is a separate thing.

Reviewer: I think somewhere in here we want to tell the buildpack author how they can elect to clear the environment. It is by setting `clear-env` to true in buildpack.toml: https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml

Reviewer: We should also mention, for buildpacks that DO allow themselves to be configured by the end user, there is a special convention for that:

https://github.com/buildpacks/spec/blob/main/buildpack.md#provided-by-the-platform

i.e., `BP_*`, and `BPL_*`. Just as an example, here is a Paketo buildpack that exposes 3 configuration variables that all have the prefix: https://github.com/paketo-buildpacks/ca-certificates?tab=readme-ov-file#configuration

<!--more-->

> Buildpack authors may elect to clear user-defined environment variables on the execution of `bin/detect` and `bin/build`. This is achievable by setting `clear-env` to `true` in [buildpack.toml](https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml); by default `clear-env` is set to `false`.

Reviewer: ```suggestion
Buildpack authors may elect to clear user-defined environment variables when `bin/detect` and `bin/build` are executed. This is achieved by setting `clear-env` to `true` in [buildpack.toml](https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml); by default `clear-env` is set to `false`.
```

Author: updated

> The `lifecycle` MUST also provide as environment variables any operator-provided files in `<build-config>/env` with environment variable names and contents matching the file names and contents. This applies for all values of `clear-env` or if `clear-env` is undefined in `buildpack.toml`.

Reviewer: I think this bit of detail if maybe a little confusing in the context. We can just say something like 

> Note that buildpacks CANNOT declare "I am not configurable by the platform". That is a separate thing.

above (but more professionally 🙃 )

Author: added

Author: added

> * When `clear-env` is set to `true` for a given buildpack, the `lifecycle` MUST NOT set user-provided environment variables in `/bin/detect` or `/bin/build`.

Reviewer: ```suggestion
* When `clear-env` is set to `true` for a given buildpack, the `lifecycle` will not set user-provided environment variables when running `/bin/detect` or `/bin/build` for the given buildpack.
```

Author: updated

> * When `clear-env` in `buildpack.toml` is set to `true` for a given buildpack, the lifecycle MUST NOT set user-provided environment variables in the environment of `/bin/detect` or `/bin/build`.

Reviewer: Can we re-word this to make it a little more user friendly? This is one of those places where we're talking to a lifecycle implementor vs a buildpack author

Author: done

Author: done

> * There is a special convention to configure Buildpacks that allow end user customization; shown in the following table.

Reviewer: ```suggestion
* If a buildpack does allow customization by the end-user through the environment (`clear-env` is `false`), there is a special convention for naming the available environment variables, shown in the following table:
```

Author: updated

> The following environment variables MUST NOT be overridden by the `lifecycle`

Reviewer: ```suggestion
```

Author: updated

>   * For layer path environment variables, user-provided values are prepended before any existing values and are delimited by the OS path list separator.

Reviewer: Let's move this stuff to the "specify env" section. It's not super relevant for clear env.

Author: done

| Env Variable           | Description                                       | Detect | Build | Launch |
|------------------------|---------------------------------------------------|--------|-------|--------|
| `BP_*`                 | User-provided variable for buildpack              | [x]    | [x]   |        |
| `BPL_*`                | User-provided variable for exec.d                 |        |       | [x]    |

> | `HOME`                 | Current user's home directory                     | [x]    | [x]   | [x]    |

Reviewer: ```suggestion
```

Author: updated

> The following environment variables MUST NOT be overridden by the `lifecycle`

Reviewer: Same with this section

Author: done

### Further Reading

> For more about environment variables, see the [customize buildpack behavior with build-time environment variables](https://buildpacks.io/docs/for-app-developers/how-to/build-inputs/configure-build-time-environment/) documentation and the [Environment](https://github.com/buildpacks/spec/blob/main/buildpack.md#environment) specification.

Reviewer: ```suggestion
For more about how environment variables are specified by end-users, see the page for how to [customize buildpack behavior with build-time environment variables](/docs/for-app-developers/how-to/build-inputs/configure-build-time-environment/).
```

Author: updated
